spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java

```diff
@@ -16,125 +16,88 @@
-		Producer<T> producer = this.producerCache.get(schemaTopic);
-		if (producer == null) {
-			producer = this.pulsarProducerFactory.createProducer(schema);
-			this.producerCache.put(schemaTopic, producer);
+		catch (Exception ex) {
+			throw PulsarClientException.unwrap(ex);
 		}
-		return producer.sendAsync(message);
-	}
-
-	public CompletableFuture<MessageId> sendAsync(T message, MessageRouter messageRouter) throws PulsarClientException {
-		final Schema<T> schema = this.schema != null ? this.schema : SchemaUtils.getSchema(message);
-		final SchemaTopic schemaTopic = getSchemaTopic(schema, this.pulsarProducerFactory, messageRouter);
-		Producer<T> producer = this.producerCache.get(schemaTopic);
-		if (producer == null) {
-			producer = this.pulsarProducerFactory.createProducer(schema, messageRouter);
-			this.producerCache.put(schemaTopic, producer);
-		}
-		return producer.sendAsync(message);
-	}
-
-	private SchemaTopic getSchemaTopic(Schema<T> schema, PulsarProducerFactory<T> pulsarProducerFactory, MessageRouter messageRouter) {
-		return new SchemaTopic(schema, (String) pulsarProducerFactory.getProducerConfig().get("topicName"), messageRouter);
 	}
 
-	public void setDefaultTopicName(String defaultTopicName) {
-		this.defaultTopicName = defaultTopicName;
-		this.pulsarProducerFactory.getProducerConfig().put("topicName", defaultTopicName);
+	@Override
+	public CompletableFuture<MessageId> sendAsync(String topic, T message, MessageRouter messageRouter) throws PulsarClientException {
+		final String topicName = resolveTopicName(topic);
+		this.logger.trace(() -> String.format("Sending msg to '%s' topic", topicName));
+		final Producer<T> producer = prepareProducerForSend(topic, message, messageRouter);
+		return producer.sendAsync(message)
+				.whenComplete((msgId, ex) -> {
+					if (ex == null) {
+						this.logger.trace(() -> String.format("Sent msg to '%s' topic", topicName));
+						// TODO success metrics
+					}
+					else {
+						this.logger.error(ex, () -> String.format("Failed to send msg to '%s' topic", topicName));
+						// TODO fail metrics
+					}
+					closeProducerAsync(producer);
+				});
 	}
 
-	public Schema<T> getSchema() {
-		return this.schema;
+	private String resolveTopicName(String userSpecifiedTopic) {
+		if (StringUtils.hasText(userSpecifiedTopic)) {
+			return userSpecifiedTopic;
+		}
+		return Optional.ofNullable(this.producerFactory.getProducerConfig().get("topicName"))
+				.map(Object::toString)
+				.orElseThrow(() -> new IllegalArgumentException("Topic must be specified when no default topic is configured"));
 	}
 
-	public void setSchema(Schema<T> schema) {
-		this.schema = schema;
+	private Producer<T> prepareProducerForSend(String topic, T message, MessageRouter messageRouter) throws PulsarClientException {
+		Schema<T> schema = SchemaUtils.getSchema(message);
+		return this.producerFactory.createProducer(topic, schema, messageRouter);
 	}
 
-	private class SchemaTopic {
```

Comment:
We will bring this back when/if we bring back producer caching. I would be in favor of putting the caching in a CachingProducerFactory impl.